### DISCUSSION TITLE
Help with scenes

### DISCUSSION THREAD

**Discussion by kleiton** [Amsterdam-3000/ProstavaBot#5] [Q&A] [answered]
Hello my friends, 

I would like to thank you for making the code of your chatbot project available on github, I am studying for it and learning a lot!
If you allow me, I would like to ask a question about the scenes part.
I'm trying to create a scene with Scenes.WizardScene and use in the same structure that you created, but I'm having a problem with the types of the fields,

I made a copy of your repository, changed some files and tried to use it, if you can take a look to indicate where the error is, I would appreciate it immensely
https://github.com/kleiton/ProstavaBot_example/blob/main/src/scenes/test.ts

Something like this:

import { Composer, Context, Markup, Scenes, session, Telegraf } from "telegraf";
import { SCENE } from "../commons/constants";
// import { UpdateContext } from "../commons/interfaces";
import { ConsultarPdContext } from "../middlewares";

```
const step0 = async (ctx) => {
    ctx.replyWithMarkdown(`0`);
    return ctx.wizard.next();
};

const step1 = new Composer<ConsultarPdContext>();
step1.hears(/^200[8-9]|201[\d]|202[0-1]$/, async (ctx) => {
    await ctx.replyWithMarkdown(`1`);
    return ctx.wizard.next();
});

const step2 = new Composer<ConsultarPdContext>();
step2.hears(/\d{6} - [\w].*$/, async (ctx) => {
    await ctx.replyWithMarkdown(`2`);
    return ctx.wizard.next();
});

const step3 = new Composer<ConsultarPdContext>();
step3.hears(/^\d{1,6}$/, async (ctx) => {
    await ctx.replyWithMarkdown(`3`);
    return ctx.wizard.next();
});

const step4 = new Composer<ConsultarPdContext>();
step4.hears(/\w|\d/, async (ctx) => {
    await ctx.replyWithMarkdown(`4
  
  ${ctx.session.consultarPd.exercicioPd}
  ${ctx.session.consultarPd.ugPd}
  ${ctx.session.consultarPd.codigoPd}`);
    return ctx.scene.leave();
});

**export const testScene** = new Scenes.**WizardScene**(SCENE.TEST, (ctx) => step0(ctx), step1, step2, step3, step4);
```

And use that scene in the file Scenes/**stage**.ts
```
import { Scenes } from "telegraf";
import { startScene, helpScene, settingsScene, **testScene** } from ".";
import { UpdateContext } from "../commons/interfaces";

//Create indefinite stage for all scenes
export const mainStage = new Scenes.Stage<UpdateContext>([startScene, helpScene, settingsScene, testScene]);
```

I get this error

Type 'WizardScene<ConsultarPdContext>' is not assignable to type 'BaseScene<UpdateContext>'.
  Types of property 'enterHandler' are incompatible.
    Type 'MiddlewareFn<ConsultarPdContext>' is not assignable to type 'MiddlewareFn<UpdateContext>'.
      Type 'UpdateContext' is missing the following properties from type 'ConsultarPdContext': myContextProp, session, wizardts(2322)

I tried to change the type of the function but without success too

`export const mainStage = new Scenes.WizardScene<UpdateContext>([startScene, helpScene, settingsScene, testScene]);
`
Thank you!!

**Comment by usebooz** [comment, marked as answer]
Hello! 

Thanks for your thanks.

You can't use scenes with different contexts. Create all scenes with your context and finally use this context for stage creation. 

For instance:

**Scenes/start.ts**
```typescript
import { Scenes } from "telegraf";

import { SCENE } from "../commons/constants";
import { ConsultarPdContext } from "../middlewares"

export const startScene = new Scenes.BaseScene<ConsultarPdContext>(SCENE.START);

//TODO controllers
//TEST
startScene.enter((ctx) => ctx.reply("Hi"));
startScene.leave((ctx) => ctx.reply("Bye"));
```

**Scenes/stage.ts**
```typescript
import { Scenes } from "telegraf";
import { startScene, helpScene, settingsScene, testScene } from ".";
import { ConsultarPdContext } from "../middlewares";

//Create indefinite stage for all scenes
export const mainStage = new Scenes.Stage<ConsultarPdContext>([startScene, testScene]);
```

**Reply by kleiton** [reply]
Hello my friend, thank you very much for your kindness in replying, I understood the problem you pointed out and now I am doing it that way.
Please forgive me for the long code

I created the interface in the following way:
**src/commons/interfaces.ts**

```
import { Scenes, Context } from "telegraf";
import { I18nContext } from "@edjopato/telegraf-i18n";

export interface UpdateContext extends Context {
    readonly i18n: I18nContext;
    scene: Scenes.SceneContextScene<UpdateContext>;
}

interface cadastro {
    name: string;
    phone?: string;
    email?: string;
    password: string;
}

interface consultarPd {
    exercicioPd: string;
    ugPd: string;
    codigoPd: string;
    credorPd?: string | string;
    ambienteConsulta?: string;
}

interface CadastroSession extends Scenes.WizardSession {
    cadastro: cadastro;
}
export interface CadastroContext extends Context {
    myContextProp: string;
    session: CadastroSession;
    scene: Scenes.SceneContextScene<CadastroContext, Scenes.WizardSessionData>;
    wizard: Scenes.WizardContextWizard<CadastroContext>;
}

interface ConsultarPdSession extends Scenes.WizardSession {
    consultarPd: consultarPd;
}
export interface ConsultarPdContext extends Context {
    myContextProp: string;
    session: ConsultarPdSession;
    scene: Scenes.SceneContextScene<ConsultarPdContext, Scenes.WizardSessionData>;
    wizard: Scenes.WizardContextWizard<ConsultarPdContext>;
}
```

I use this context in the scene as follows:
**src/scenes/test.ts**

```
import { Composer, Context, Markup, Scenes, session, Telegraf } from "telegraf";
import { SCENE } from "../commons/constants";
import { ConsultarPdContext } from "../middlewares";

const step0 = async (ctx) => {
    ctx.replyWithMarkdown(`0`);
    return ctx.wizard.next();
};

const step1 = new Composer<ConsultarPdContext>();
step1.hears(/^200[8-9]|201[\d]|202[0-1]$/, async (ctx) => {
    await ctx.replyWithMarkdown(`1`);
    return ctx.wizard.next();
});

const step2 = new Composer<ConsultarPdContext>();
step2.hears(/\d{6} - [\w].*$/, async (ctx) => {
    await ctx.replyWithMarkdown(`2`);
    return ctx.wizard.next();
});

const step3 = new Composer<ConsultarPdContext>();
step3.hears(/^\d{1,6}$/, async (ctx) => {
    await ctx.replyWithMarkdown(`3`);
    return ctx.wizard.next();
});

const step4 = new Composer<ConsultarPdContext>();
step4.hears(/\w|\d/, async (ctx) => {
    await ctx.replyWithMarkdown(`4
  
  ${ctx.session.consultarPd.exercicioPd}
  ${ctx.session.consultarPd.ugPd}
  ${ctx.session.consultarPd.codigoPd}`);
    return ctx.scene.leave();
});

export const testScene = new Scenes.WizardScene(SCENE.TEST, (ctx) => step0(ctx), step1, step2, step3, step4);
```

On the Scenes's stage I refer to Scena passing the same context
**src/scenes/stage.ts**

```
import { Scenes } from "telegraf";
import { testScene } from ".";
import { ConsultarPdContext } from "../commons/interfaces";
export const mainStage2 = new Scenes.Stage<ConsultarPdContext>([testScene]);
```

On the MIddleware's stage I refer to mainStage2
**src/middlewares/stage.ts**

```
import { mainStage2 } from "../scenes";
export const addAllScenes = mainStage2.middleware();
```

and finally, I pass the addAllScenes middleware to the bot in the index
**src/index.ts**

```
import { bot } from "./commons/bot";
import { SCENE } from "./commons/constants";
import {addAllScenes } from "./middlewares";

bot.use(addAllScenes);
```

I'm having a problem with the index only, I get the following error:

> (alias) const addAllScenes: MiddlewareFn<ConsultarPdContext>
> import addAllScenes
> Argument of type 'MiddlewareFn<ConsultarPdContext>' is not assignable to parameter of type 'Middleware<UpdateContext>'.
>   Type 'MiddlewareFn<ConsultarPdContext>' is not assignable to type 'MiddlewareFn<UpdateContext>'.
>     Type 'UpdateContext' is missing the following properties from type 'ConsultarPdContext': myContextProp, session, wizardts(2345)



I saw that the middleware() function receives a BaseScene
**src/middlewares/stage.ts**

But I was unable to change to work with a WizardScene

**node_modules/telegraf/src/scenes/stage.ts**

```
  middleware() {
    const handler = Composer.compose<C>([
      (ctx, next) => {
        const scenes: Map<string, BaseScene<C>> = this.scenes
        const scene = new SceneContextScene<C, D>(ctx, scenes, this.options)
        ctx.scene = scene
        return next()
      },
      super.middleware(),
      Composer.lazy<C>((ctx) => ctx.scene.current ?? Composer.passThru()),
    ])
    return Composer.optional(isSessionContext, handler)
  }
```

can you guide me how to proceed, please?

thank you so much!

**Reply by usebooz** [reply]
You need to read more about [TypeScript generics](https://www.typescriptlang.org/docs/handbook/generics.html).

Replace in all your project old context UpdateContext with ConsultarPdContext. 

For instance:
**commons/bot.ts**
```typescript
import { Telegraf } from "telegraf";
import { ConsultarPdContext } from "../middlewares"
import { CONFIG } from "./config";

//Create bot instance
export const bot = new Telegraf< ConsultarPdContext >(CONFIG.TELEGRAM_TOKEN);
```

And etc. 

My pleasure

**Reply by kleiton** [reply]
I got it, I adjusted it but there is still a problem, anyway I will read the Generics material that you informed

Thank you very much for your patience in responding!
